drivers/lg/displays/ls5.cr

```diff
@@ -0,0 +1,291 @@
+  )
+  mute_video(state) if layer.video? || layer.audio_video?
+  mute_audio(state) if layer.audio? || layer.audio_video?
+  end
+
+  def mute_video(state : Bool = true)
+    state = state ? 1 : 0
+    do_send(Command::ScreenMute, state, name: "mute_video")
+  end
+
+  def mute_audio(state : Bool = true)
+    # Do nothing if already in desired state
+    return if (self[:audio_mute]?.try &.as_bool) == state
+    state = state ? 0 : 1
+    do_send(Command::VolumeMute, state, name: "mute_audio")
+  end
+
+  enum Ratio
+    Square  = 0x01
+    Wide    = 0x02
+    Zoom    = 0x04
+    Scan    = 0x09
+    Program = 0x06
+  end
+  def aspect_ratio(ratio : Ratio)
+    do_send(Command::AspectRatio, ratio.value, name: "aspect_ratio", delay: 1.second)
+  end
+
+  def do_poll
+    if @rs232
+      power?
+      if self[:hard_power]?.try &.as_bool
+        screen_mute?
+        input?
+        volume_mute?
+        volume?
+      end
+    elsif self[:connected]?.try &.as_bool
+      screen_mute?
+
+      if @id_num == 1
+        input?
+        volume_mute?
+        volume?
+      end
+    elsif self[:power_target]?.try &.as_bool
+      power(true)
+    end
+  end
+
+  def input?(priority : Int32 = 0)
+    do_send(Command::Input, 0xFF, 'x', priority: priority)
+  end
+
+  {% for name in ["Volume", "Contrast", "Brightness", "Sharpness"] %}
+    @[Security(Level::Administrator)]
+    def {{name.id.downcase}}(value : Int32)
+      val = value.clamp(0, 100)
+      do_send(Command::{{name.id}}, val, name: {{name.id.downcase.stringify}})
+    end
+  {% end %}
+
+  def pm_mode(mode : Int32 = 3)
+    do_send(Command::PmMode, mode, 's', name: "pm_mode")
+  end
+
+  # 0 = Off, 1 = lock all except Power buttons, 2 = lock all buttons. Default to 2 as power off from local button results in network offline
+  def local_button_lock(state : Bool = true)
+    val = state ? 2 : 0
+    do_send(Command::LocalButtonLock, val, 't', name: "local_button_lock")
+  end
+
+  def no_signal_off(state : Bool = false)
+    val = state ? 1 : 0
+    do_send(Command::NoSignalOff, val, 'f', name: "disable_no_signal_off")
+  end
+
+  def auto_off(state : Bool = false)
+    val = state ? 1 : 0
+    do_send(Command::AutoOff, val, 'm', name: "disable_auto_off")
+  end
+
+  def wake_on_lan(state : Bool = true)
+    val = state ? 1 : 0
+    do_send(Command::WakeOnLan, val, 'f', name: "enable_wake_on_lan")
+  end
+
+  def wake(broadcast : String? = nil)
+    if mac = setting?(String, :mac_address)
+      # config is the database model representing this device
+      wake_device(mac, broadcast)
+      logger.debug {
+        info = "Wake on Lan for MAC #{mac}"
+        if b = broadcast
+          info += " directed to VLAN #{b}"
+        end
+        info
+      }
+    else
+      logger.debug { "No MAC address provided" }
+    end
+  end
+
+  DUPLICATES = ["Contrast", "NoSignalOff", "AutoOff", "PmMode"]
+  def received(data, task)
+    command = Command.from_value(data[0]).to_s
```

Comment:
It will be much safer to work with this as it's enum value rather than in string form.

Reply:
Yeah I tried using the enum values for this but because some of the commands use the same letter/value, then I can't actually distinguish between their responses alone from the first character in the response string.

Pretty much if there's duplicate values in an enum, it'll always default to the one defined first so AutoOff in this case. Let me know if you have a better idea of dealing with this https://play.crystal-lang.org/#/r/9tt5

Also, this comments try to explain that https://github.com/PlaceOS/drivers/blob/3dea9fef83f703d0af9ef10443dda86bf851bfd9/drivers/lg/displays/ls5.cr#L223-L225

Reply:
Ah I see. Looking at the protocol, commands are two bytes. Rather than fudging things on the parsing, it be more reliable to rework the way this is modelled to capture this.

One approach that could work is to pack the two command bytes (`UInt8`'s) into a single `UInt16`. This will then let the commands be represented, without ambiguity and also ensure that everything needed to serialise and deserialise them is define clearly in one spot.

```crystal
enum Command : UInt16
  Power    = 0x6b61 # k a
  SetInput = 0x7962 # x b
  # etc
  
  def self.new(data1, data2)
    self.new((data1.to_u16 << 8) ^ data2.to_u8)
  end
end

p! Command.new('k'.ord, 'a'.ord)
```
https://play.crystal-lang.org/#/r/9tvu

The only slightly annoying thing is I'm not aware of a way to get `Char#ord` values from within a macro or force compile time eval of these to make the actual value declarations a little neater. E.g. ideally this would be:

```crystal
enum Command : UInt16
  Power = bitpack('k', 'a')
  # ...
end
```

Tagging @stakach and @caspiano for other thoughts on this one too as defining a clean way to express multi-byte Enum's is a common problem worth solving.

Reply:
P.S. see crystal-lang/crystal#9830 for a proposal to solve this at the language level. Hopefully it's generally useful, however if there's no interest there can look at adding something to the driver framework.

Reply:
Hey @KimBurgess , if I do switch over to packing 2 bytes into a command like `NoSignalOff    = 0x6667 # fg`, I'm not sure how I'll be able to use the enum to parse responses.

So the command for `no_signal_off(false)` is `"fg 01 00\r"` and the response is `"g 01 OK00x"`. Since Command::NoSignalOff is representative of the two bytes which are only present in the command that's sent and the response only contains "g" and not "fg", this will make it difficult if not impossible to figure out what command the response is actually from.

Reply:
Yep - I see what you mean. In cases like this where the protocol does not provide the ability to uniquely distinguish the response from the response data alone, you may want to define response parsing as a callback block, in place of the single `received` method.

Take a look at ["send and callback" here](https://github.com/PlaceOS/drivers/blob/master/docs/writing-a-driver.md#streaming-io).